edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/EdgeletModule.cs

```diff
@@ -84,14 +84,25 @@ protected override void Load(ContainerBuilder builder)
         {
             // IModuleClientProvider
             builder.Register(
-                    c => new ModuleClientProvider(
-                        c.Resolve<ISdkModuleClientProvider>(),
-                        this.upstreamProtocol,
-                        this.proxy,
-                        this.productInfo,
-                        this.closeOnIdleTimeout,
-                        this.idleTimeout,
-                        this.useServerHeartbeat))
+                    c =>
+                    {
+                        IModuleManager m = c.Resolve<IModuleManager>();
+
+                        // NOTE: Deadlock risk
+                        SystemInfo system = m.GetSystemInfoAsync(CancellationToken.None)
+                            .ConfigureAwait(false)
+                            .GetAwaiter()
+                            .GetResult();
+
+                        return new ModuleClientProvider(
```

Comment:
Can we instead pass the IModule manager to ```ModuleClientProvider``` and have it call GetSystemInfoAsync during module creation?

Reply:
Yes, that is a safer idea than what I have now.

Reply:
8cc6a8abf9daf8db5f0bdab01e3587c4da004077